[c10d] Add a collective time estimator for NCCL comms

fduwjj opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #149343

We want to upstream the feature from new nccl for users to estimate comm time.

Resolves #147753

cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @wz337 @wconstab @d4l3k @c-p-i-o

fduwjj committed:
[RFC] Add a collective time estimator for NCCL comms

[ghstack-poisoned]

kwen2501 commented on torch/csrc/distributed/c10d/Backend.hpp:
```diff
@@ -73,6 +73,10 @@ class TORCH_API Backend : public torch::CustomClassHolder {
     return false;
   }
 
+  virtual bool supportsCommTimeEstimation() const {
```
`supportsTimeEstimation` for less letters

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -3249,6 +3249,25 @@ c10::intrusive_ptr<Work> ProcessGroupNCCL::endCoalescing() {
+
+float ProcessGroupNCCL::endCommTimeEstimate() {
+#ifdef NCCL_SIM_INFO_INITIALIZER
+  auto simInfo = std::make_shared<ncclSimInfo_t>();
```
why using a shared_ptr when we have no intention to share? (`simInfo` dies within this function's scope).
At `ncclCollectiveEstimateEnd`, I wonder if we can directly return the time.

kwen2501 commented on torch/csrc/distributed/c10d/NCCLUtils.cpp:
```diff
@@ -118,6 +118,12 @@ std::shared_ptr<NCCLComm> NCCLComm::create_scalable(
 #endif // NCCL_HAS_INIT_RANK_SCALABLE
 #endif // NCCL_HAS_CONFIG
 
+#ifdef NCCL_SIM_INFO_INITIALIZER
+ncclResult_t NCCLComm::ncclCollectiveEstimateEnd(ncclSimInfo_t* simInfo) {
+  return ncclGroupSimulateEnd(simInfo);
+}
+#endif // NCCL_SIM_INFO_INITIALIZER
+
```
If it makes the usage easier, maybe we can make this function directly return the time?

kwen2501 reviewed (commented):
LGTM

fduwjj committed:
Update on "[RFC] Add a collective time estimator for NCCL comms"


Not for landing and no unit test right now. Just use this PR to collective feedback to see if this is something we want to do.

cc H-Huang awgu kwen2501 wanchaol fegin wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

fduwjj committed:
Update on "[RFC] Add a collective time estimator for NCCL comms"


Not for landing and no unit test right now. Just use this PR to collective feedback to see if this is something we want to do.

cc H-Huang awgu kwen2501 wanchaol fegin wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d] Add a collective time estimator for NCCL comms"


We want to upstream the feature from new nccl for users to estimate comm time.

cc H-Huang awgu kwen2501 wanchaol fegin wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

awgu commented on torch/distributed/distributed_c10d.py:
```diff
@@ -2606,6 +2606,46 @@ def _coalescing_manager(
+    device: Optional[torch.device] = None,
+):
+    """
+    Context manager used to coalesce collectives or P2P operations when possible.
```
seems accidentally copied from `_coalescing_manager`

fduwjj committed:
Update on "[c10d] Add a collective time estimator for NCCL comms"


We want to upstream the feature from new nccl for users to estimate comm time.

cc H-Huang awgu kwen2501 wanchaol fegin wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

fduwjj commented:
unit test passed and now it is ready for reviewing.

fduwjj committed:
Update on "[c10d] Add a collective time estimator for NCCL comms"


We want to upstream the feature from new nccl for users to estimate comm time.

cc H-Huang awgu kwen2501 wanchaol fegin wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

kwen2501 commented on torch/distributed/distributed_c10d.py:
```diff
@@ -2606,6 +2606,50 @@ def _coalescing_manager(
+
+
+@contextlib.contextmanager
+def _collective_estimator(
```
nit: `_time_estimator`

kwen2501 commented on torch/distributed/distributed_c10d.py:
```diff
@@ -2606,6 +2606,50 @@ def _coalescing_manager(
+        >>> # Synchronous ops
+        >>> with _collective_estimator() as cm:
+        >>>     for i in range(num_colls):
+        >>>         dist.all_reduce(tensors[i])
```
(1) is required because only with a real communicator can we do accurate estimation. The communicator internally has knowledge about the links it runs on (e.g. NCCL knows whether we are intra-node or inter-node, whether the links are NVLink or PCI-e or IB).

(2) is similar to CUDA Graph capturing -- in the capturing context, nothing is actually run. But yeah, we should document the expectation well.

kwen2501 replied:
That said, I don't know if Inductor uses real tensor or fake tensor for scheduling, we should try to support whatever Inductor's using.

fduwjj replied:
I can add a TODO for (2) to support what inductor is using as well. Any customer from the compiler side is interested to try this tool?

kwen2501 commented on torch/distributed/__init__.py:
```diff
@@ -129,6 +129,7 @@ def breakpoint(rank: int = 0, skip: int = 0):
         _all_gather_base,
         _coalescing_manager,
         _CoalescingManager,
+        _collective_estimator,
```
nit: `_time_estimator`

kwen2501 commented on torch/csrc/distributed/c10d/NCCLUtils.cpp:
```diff
@@ -118,6 +118,14 @@ std::shared_ptr<NCCLComm> NCCLComm::create_scalable(
 #endif // NCCL_HAS_INIT_RANK_SCALABLE
 #endif // NCCL_HAS_CONFIG
 
+#ifdef NCCL_SIM_INFO_INITIALIZER
+float NCCLComm::ncclCollectiveEstimateEnd(
+    std::unique_ptr<ncclSimInfo_t> simInfo) {
+  C10D_NCCL_CHECK(ncclGroupSimulateEnd(simInfo.get()), std::nullopt);
+  return simInfo->estimatedTime;
+}
+#endif // NCCL_SIM_INFO_INITIALIZER
+
```
```
float NCCLComm::ncclCollectiveEstimateEnd() {
  ncclSimInfo_t simInfo = NCCL_SIM_INFO_INITIALIZER;
  C10D_NCCL_CHECK(ncclGroupSimulateEnd(&simInfo), std::nullopt);
  return simInfo->estimatedTime;
}
```
so that this API's signature does not depend on `ncclSimInfo_t`

kwen2501 replied:
Actually, since the implementation here does not use the comm, you can put `ncclGroupSimulateEnd` directly in `ProcessGroupNCCL` instead of `NCCLComm`.

kwen2501 commented on torch/csrc/distributed/c10d/NCCLUtils.hpp:
```diff
@@ -234,6 +234,11 @@ class NCCLComm {
       std::vector<uint64_t>& ranks_ull);
 #endif // NCCL_HAS_COMM_SPLIT
 
+#ifdef NCCL_SIM_INFO_INITIALIZER
+  static float ncclCollectiveEstimateEnd(
+      std::unique_ptr<ncclSimInfo_t> simInfo);
+#endif // NCCL_SIM_INFO_INITIALIZER
+
```
We can remove the `#ifdef` if we remove `ncclSimInfo_t` from this signature. See my other comment in cpp.

Minor name pref: `endSimulate`

kwen2501 commented on torch/distributed/distributed_c10d.py:
```diff
@@ -2606,6 +2606,50 @@ def _coalescing_manager(
     # Otherwise, the backend has sync'ed at CPP level
 
 
+class _CollectiveEstimator:
```
nit: `_time_estimator`

kwen2501 reviewed (approved):
LGTM. Thanks

fduwjj committed:
Update on "[c10d] Add a collective time estimator for NCCL comms"


We want to upstream the feature from new nccl for users to estimate comm time.

cc H-Huang awgu kwen2501 wanchaol fegin wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d] Add a collective time estimator for NCCL comms"


We want to upstream the feature from new nccl for users to estimate comm time.

cc H-Huang awgu kwen2501 wanchaol fegin wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

fduwjj commented:
@pytorchbot merge
